opentelemetry/proto/logs/v1/logs.proto

```diff
@@ -0,0 +1,130 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package opentelemetry.proto.logs.v1;
+
+import "opentelemetry/proto/common/v1/common.proto";
+import "opentelemetry/proto/resource/v1/resource.proto";
+
+option java_multiple_files = true;
+option java_package = "io.opentelemetry.proto.logs.v1";
+option java_outer_classname = "LogsProto";
+option go_package = "github.com/open-telemetry/opentelemetry-proto/gen/go/logs/v1";
+
+// A collection of InstrumentationLibraryLogs from a Resource.
+message ResourceLogs {
+    // The resource for the logs in this message.
+    // If this field is not set then no resource info is known.
+    opentelemetry.proto.resource.v1.Resource resource = 1;
+
+    // A list of InstrumentationLibraryLogs that originate from a resource.
+    repeated InstrumentationLibraryLogs instrumentation_library_logs = 2;
+}
+
+// A collection of Logs produced by an InstrumentationLibrary.
+message InstrumentationLibraryLogs {
+    // The instrumentation library information for the logs in this message.
+    // If this field is not set then no library info is known.
+    opentelemetry.proto.common.v1.InstrumentationLibrary instrumentation_library = 1;
+
+    // A list of log records.
+    repeated LogRecord logs = 2;
+}
+
+// Possible values for LogRecord.SeverityNumber.
+enum SeverityNumber {
+    UNDEFINED_SEVERITY_NUMBER = 0;
+    TRACE  = 1;
+    TRACE2 = 2;
+    TRACE3 = 3;
+    TRACE4 = 4;
+    DEBUG  = 5;
+    DEBUG2 = 6;
+    DEBUG3 = 7;
+    DEBUG4 = 8;
+    INFO   = 9;
+    INFO2  = 10;
+    INFO3  = 11;
+    INFO4  = 12;
+    WARN   = 13;
+    WARN2  = 14;
+    WARN3  = 15;
+    WARN4  = 16;
+    ERROR  = 17;
+    ERROR2 = 18;
+    ERROR3 = 19;
+    ERROR4 = 20;
+    FATAL  = 21;
+    FATAL2 = 22;
+    FATAL3 = 23;
+    FATAL4 = 24;
+}
+
+// Masks for LogRecord.flags field.
+enum LogRecordFlags {
+    UNDEFINED_LOG_RECORD_FLAG = 0;
+    TRACE_FLAGS_MASK = 0x000000FF;
+}
+
+// A log record according to OpenTelemetry Log Data Model:
+// https://github.com/open-telemetry/oteps/blob/master/text/logs/0097-log-data-model.md
+message LogRecord {
+    // time_unix_nano is the time when the event occurred.
+    // Value is UNIX Epoch time in nanoseconds since 00:00:00 UTC on 1 January 1970.
+    // Value of 0 indicates unknown or missing timestamp.
+    fixed64 time_unix_nano = 1;
+
+    // A unique identifier for a trace. All logs from the same trace share
+    // the same `trace_id`. The ID is a 16-byte array. An ID with all zeroes
+    // is considered invalid. Can be set for logs that are part of request processing
+    // and have an assigned trace id. Optional.
+    bytes trace_id = 2;
```

Comment:
Is there a reason for having trace-related fields in the most prominent place of the struct? I would prefer the actually log-related fields to be at the top.

Reply:
I followed the ordering with have in log data model OTEP: https://github.com/open-telemetry/oteps/blob/master/text/logs/0097-log-data-model.md

Reply:
I would suggest moving trace context fields to the bottom of the struct. It doesn't even apply to large classes of logs.

Reply:
I am a bit reluctant because of 2 reasons:
- It will be a deviation from the order we have in the data model.
- It will result in worse packing of the struct and increased memory usage due to alignment rules in certain languages (e.g. in Go, which probably matters most from memory usage perspective since it is used by Collector).

Reply:
Since these are byte arrays of unknown (to proto) length, why would they impact packing of struct?

Which data model are you referring to? Given that this is all new dev, I don't really buy the "legacy" argument. If I were a newcomer to OTEL logging, I would be confused by seeing trace fields given higher ranking than the actual log fields.

Reply:
> Since these are byte arrays of unknown (to proto) length, why would they impact packing of struct?
> 

There is no problem with moving `trace_id` and `span_id`. I can move them without impacting the packing (byte arrays are slices and have 64 bit alignment in Go 64-bit arch). `flags` is the problem because it is 32 bit. Currently it is tightly packed because it is followed by another 32 bit field, which means we have no wasted space. If I move the `trace_id`/`span_id`/`flags` to the end, we will have less tight packing.

However, I thought a bit more and I can do it slightly differently and still have good packing.

I can either keep `flags` where it is and only move `trace_id`/`span_id` to the end or simply change the order of these 3 fields to `flags`/`trace_id`/`span_id` and move all 3 to the end. This way 32 bit `flags` will pack tightly with preceding `dropped_attributes_count` which is also 32 bit.

If you are OK with either of these approaches I can do it.

> Which data model are you referring to? Given that this is all new dev, I don't really buy the "legacy" argument. If I were a newcomer to OTEL logging, I would be confused by seeing trace fields given higher ranking than the actual log fields.

I was referring our OpenTelemetry [log data model](https://github.com/open-telemetry/oteps/blob/master/text/logs/0097-log-data-model.md). This is a minor argument. I am fine to reorder if you think it looks bad.

Reply:
I'm ok with flags to be above trace_id.

Reply:
Done.